### PR TITLE
Implementing filling power snapshot balances

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to #660

### CODE REVIEW

File: src/repositories/powerSnapshotRepository.ts [hunk 1]
```diff
@@ -42,14 +41,17 @@ export const getPowerBoostingSnapshotWithoutBalance = async (
...
-  Pick<
-    PowerSnapshot & PowerBoostingSnapshot,
-    'userId' | 'powerSnapshotId' | 'blockNumber'
-  >[]
+  {
+    userId: number;
+    powerSnapshotId: number;
+    blockNumber: number;
+    walletAddress: string;
+  }[]
```
Comment (mohammadranjbarz) [inline src/repositories/powerSnapshotRepository.ts:49]: I think using ```{
    userId: number;
    powerSnapshotId: number;
    blockNumber: number;
    walletAddress: string;
  }```
  is more straighforward besides we need all field to be required in previous was some of them were optional

File: src/adapters/givpowerSubgraph/givPowerSubgraphAdapter.ts [hunk 2]
```diff
@@ -1,138 +1,63 @@
...
-      user { id }
+const _queryBuilder = (params: {
+  addresses: Set<string>;
+  blockNumber: number;
+  unipoolContractId: string;
+}): string => {
+  const { addresses, blockNumber, unipoolContractId } = params;
+  let usersIn = '[';
+  addresses.forEach(address => (usersIn += `"${address}",`));
+  usersIn += ']';
```
Comment (aminlatifi) [inline src/adapters/givpowerSubgraph/givPowerSubgraphAdapter.ts]:
```suggestion
  const usersIn = '[' + addresses.map(address => `"${address}"`).join(',') + ']';
```

File: src/entities/powerBalanceSnapshot.ts [hunk 3]
```diff
@@ -27,13 +27,13 @@ export class PowerBalanceSnapshot extends BaseEntity {
...
       powerBalanceSnapshot.powerSnapshot,
   )
-  @Column()
-  powerSnapshotId: number;
-
   @Field()
   @Column('float')
   balance: number;
 
+  @Column()
```
Comment (aminlatifi) [inline src/entities/powerBalanceSnapshot.ts:34]: Why we have moved it? it had annotation on top.

File: src/repositories/powerSnapshotRepository.ts [hunk 4]
```diff
@@ -42,14 +41,17 @@ export const getPowerBoostingSnapshotWithoutBalance = async (
...
 > => {
   return await getConnection().query(
     `
-    select "userId", "powerSnapshotId", "blockNumber" from power_boosting_snapshot as boosting
+    select "userId", "powerSnapshotId", "blockNumber","walletAddress" from power_boosting_snapshot as boosting
+    inner join public."user" as u on  "userId"= u.id
```
Comment (aminlatifi) [inline src/repositories/powerSnapshotRepository.ts]:
```suggestion
    select "userId", "powerSnapshotId", "blockNumber", "walletAddress" from power_boosting_snapshot as boosting
    inner join public."user" as "user" on  "userId"= user.id 
```

File: src/repositories/powerBalanceSnapshotRepository.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { PowerBalanceSnapshot } from '../entities/powerBalanceSnapshot';
+
+export const createPowerSnapshotBalances = async (params: {
+  powerSnapshotId: number;
+  userId: number;
+  balance: number;
+}): Promise<PowerBalanceSnapshot> => {
+  return PowerBalanceSnapshot.create(params).save();
```
Comment (aminlatifi) [inline src/repositories/powerBalanceSnapshotRepository.ts]: This `create` function has another overloading that accepts an array of entity-like objects. That maybe more useful for our usecase

File: src/services/cronJobs/fillSnapshotBalances.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        groupByBlockNumbers[pb.blockNumber].push(pb);
+      } else {
+        groupByBlockNumbers[pb.blockNumber] = [pb];
+      }
+    });
+  }
+  Object.keys(groupByBlockNumbers).forEach(key => {
+    fillSnapshotBalanceQueue.add({
+      blockNumber: Number(key),
+      data: groupByBlockNumbers[key],
```
Comment (aminlatifi) [inline src/services/cronJobs/fillSnapshotBalances.ts]: It seems there is no limit on the number of items per block! I mean we would have a huge number of balances to fetch for one block!

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/services/cronJobs/fillSnapshotBalances.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        const balances =
+          await getGivPowerSubgraphAdapter().getUserPowerBalanceInBlockNumber({
+            blockNumber,
+            walletAddresses: data.map(user =>
+              user?.walletAddress?.toLowerCase(),
+            ),
+          });
+
+        await Promise.all(
+          data.map(item => {
```
Comment (aminlatifi) [inline src/services/cronJobs/fillSnapshotBalances.ts:110]: If we change createPowerSnapshotBalances to accept an array, we won't need this Pomise all

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/entities/powerBalanceSnapshot.ts (hunk 3 above)
Comment (mohammadranjbarz) [reply src/entities/powerBalanceSnapshot.ts:34]: I just wanted to keep `powerSnapshotId` and `together` if you see the whole file, you will see it's just moving some lines

File: src/adapters/givpowerSubgraph/givPowerSubgraphAdapter.ts (hunk 2 above)
Comment (mohammadranjbarz) [reply src/adapters/givpowerSubgraph/givPowerSubgraphAdapter.ts]: Done

File: src/repositories/powerSnapshotRepository.ts (hunk 4 above)
Comment (mohammadranjbarz) [reply src/repositories/powerSnapshotRepository.ts]: I was using `user` first but I get this error
```
error: syntax error at or near "user"
    at Parser.parseErrorMessage (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/index.ts:7:48)
```

but after changing it to `u` it got fixed,  I think it's related to `user` DB

File: src/services/cronJobs/fillSnapshotBalances.ts (hunk 7 above)
Comment (mohammadranjbarz) [reply src/services/cronJobs/fillSnapshotBalances.ts:110]: Done

File: src/repositories/powerSnapshotRepository.ts (hunk 4 above)
Comment (mohammadranjbarz) [reply src/repositories/powerSnapshotRepository.ts]: I was using `user` first but I get this error
```
error: syntax error at or near "user"
    at Parser.parseErrorMessage (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/Users/renjer/Documents/src/giveth/impact-graph/node_modules/pg-protocol/src/index.ts:7:48)
```

but after changing it to `u` it got fixed,  I think it's related to `user` DB

File: src/entities/powerBalanceSnapshot.ts (hunk 3 above)
Comment (mohammadranjbarz) [reply src/entities/powerBalanceSnapshot.ts:34]: Done

File: src/services/cronJobs/fillSnapshotBalances.ts (hunk 6 above)
Comment (mohammadranjbarz) [reply src/services/cronJobs/fillSnapshotBalances.ts]: Done

File: src/services/cronJobs/fillSnapshotBalances.ts [hunk 8]
```diff
@@ -86,11 +89,13 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
...
 export function processFillPowerSnapshotJobs() {
-  logger.debug('processFillPowerSnapshotJobs() has been called');
+  logger.debug('processFillPowerSnapshotJobs() has been called ', {
+    numberOfFillPowerSnapshotBAlancesConcurrentJob,
+  });
   fillSnapshotBalanceQueue.process(
     numberOfFillPowerSnapshotBAlancesConcurrentJob,
     async (job, done) => {
-      logger.debug('processing syncUserPower job', { jobData: job.data });
+      logger.debug('processing fill powerSnapshot job', job.data);
```
Comment (aminlatifi) [inline src/services/cronJobs/fillSnapshotBalances.ts:98]: Do we need to keep these logging sections?

**Review by aminlatifi — APPROVED** [review]
Thanks @mohammadranjbarz, just that question regarding logging

**Pull request merged**